widlparser/productions.py

```diff
@@ -103,20 +103,39 @@ def _consume_semicolon(self, tokens: Tokenizer, consume_tail: bool = True) -> No
 			tokens.syntax_error(None)
 
 
-class ChildProduction(Production):
-	"""Base class for productions that have parents."""
+class ComplexProduction(Production):
+	"""
+	Base class for more complex productions.
+
+	Things that can have names, parents, etc.,
+	including all Constructs.
+	"""
 
-	parent: Optional[ChildProduction]
+	_parent: Optional[ComplexProduction]
 
-	def __init__(self, tokens: Tokenizer, parent: Optional[ChildProduction]) -> None:
+	def __init__(self, tokens: Tokenizer, parent: Optional[ComplexProduction]) -> None:
 		Production.__init__(self, tokens)
-		self.parent = parent
+		self._parent = parent
+
+	@property
+	def parent(self) -> ComplexProduction:
+		if self._parent is not None:
+			return self._parent
+		raise KeyError()
+
+	@parent.setter
+	def parent(self, parent: ComplexProduction) -> None:
```

Comment:
Should this accept an Optional[ComplexProduction]? (Trying to think is there's a use case to set it to None)

Reply:
Fine with either way. I suppose since you can construct one without, it's not an important part of the contract that once it obtains a parent it keeps one.